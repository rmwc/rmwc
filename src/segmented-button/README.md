# Segmented Button

Segmented buttons allow users to toggle the selected states of grouped buttons.

- Module **@rmwc/segmented-button**
- Import styles:
  - Using CSS Loader
    - import '@rmwc/segmented-button/styles';
  - Or include stylesheets
    - **'@material/segmented-button/dist/mdc.segmented-button.css'**
    - **'@rmwc/@rmwc/icon/icon.css'**
    - **'@material/ripple/dist/mdc.ripple.css'**
- MDC Docs: [https://material.io/develop/web/components/segmented-button/](https://material.io/develop/web/components/segmented-button/)

```jsx
<SegmentedButton label="Button" />
```

## SegmentedButton
Segmented Button

### Props

| Name | Type | Description |
|------|------|-------------|
| `children` | `React.ReactNode` | Content specified as children. |
| `danger` | `undefined \| false \| true` | Used to indicate a dangerous action. |
| `dense` | `undefined \| false \| true` | Make the Button dense. |
| `disabled` | `undefined \| false \| true` | Make the button disabled |
| `icon` | `RMWC.IconPropT` | An Icon for the Button |
| `label` | `React.ReactNode \| any` | Content specified as a label prop. |
| `outlined` | `undefined \| false \| true` | Make the button outlined. |
| `raised` | `undefined \| false \| true` | Make the Button raised. |
| `touch` | `undefined \| false \| true` | Makes the button more touch friendly. This will automatically be set true if used inside of TouchTargetWrapper. |
| `trailingIcon` | `RMWC.IconPropT` | A trailing icon for the Button |
| `unelevated` | `undefined \| false \| true` | Make the button unelevated. |


