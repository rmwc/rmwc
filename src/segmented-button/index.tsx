export * from './segment';
export * from './segmented-button';
