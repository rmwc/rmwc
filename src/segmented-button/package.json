{
  "name": "@rmwc/segmented-button",
  "version": "8.0.7",
  "description": "RMWC Segmented Button component",
  "main": "dist/index.js",
  "types": "dist/index.d.ts",
  "module": "next/index.js",
  "sideEffects": [
    "./styles.tsx",
    "./styles.js",
    "*.css"
  ],
  "publishConfig": {
    "access": "public"
  },
  "scripts": {
    "publish:dist": "npm publish dist --access=public",
    "test": "echo \"Error: no test specified\" && exit 1"
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/rmwc/rmwc.git"
  },
  "keywords": [
    "rmwc"
  ],
  "author": "Emilie Ã˜rtoft",
  "private": false,
  "license": "MIT",
  "bugs": {
    "url": "https://github.com/rmwc/rmwc/issues"
  },
  "homepage": "https://github.com/rmwc/rmwc/tree/master/src/segmented-button#readme",
  "peerDependencies": {
    "react": ">=16.8.x",
    "react-dom": ">=16.8.x"
  },
  "dependencies": {
    "@material/segmented-button": "^14.0.0",
    "@rmwc/base": "^8.0.7",
    "@rmwc/icon": "^8.0.7",
    "@rmwc/provider": "^8.0.7",
    "@rmwc/ripple": "^8.0.7",
    "@rmwc/touch-target": "^8.0.7",
    "@rmwc/types": "^8.0.7"
  },
  "gitHead": "d9befec0e3b258864b71686eecd16c16feb43cf1"
}
