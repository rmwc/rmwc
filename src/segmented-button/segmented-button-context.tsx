import React, { useContext } from 'react';

export type SegmentedButtonContextT = {
  registerSegment: (segment: any) => void;
  unregisterSegment: (segment: any) => void;
  selectType: 'multiple' | 'single';
};

export const SegmentedButtonContext =
  React.createContext<SegmentedButtonContextT>({
    registerSegment: (segment: any) => {},
    unregisterSegment: (segment: any) => {},
    selectType: 'multiple'
  });

export const SegmentedButtonProvider = ({
  value,
  children
}: {
  value: SegmentedButtonContextT;
  children: React.ReactNode;
}) => {
  return (
    <SegmentedButtonContext.Provider value={value}>
      {children}
    </SegmentedButtonContext.Provider>
  );
};

export const useSegmentedButton = () => {
  const context = useContext(SegmentedButtonContext);
  if (!context) {
    throw new Error(
      'useSegmentedButton must be used within an SegmentedButtonProvider'
    );
  }
  return context;
};
