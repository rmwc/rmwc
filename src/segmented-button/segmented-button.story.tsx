import React from 'react';

import { storiesOf } from '@storybook/react';
import { Segment, SegmentedButton } from './';

storiesOf('Segmented-Button', module).add('Segmented-Button', () => (
  <SegmentedButton>
    <Segment label="Cookies" />
  </SegmentedButton>
));
