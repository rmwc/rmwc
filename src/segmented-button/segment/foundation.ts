import React, { useEffect, useMemo, useState } from 'react';
import { emptyClientRect, useFoundation } from '@rmwc/base';
import {
  MDCSegmentedButtonSegmentFoundation,
  SegmentDetail
} from '@material/segmented-button';
import { SegmentApi, SegmentProps } from './';
import { useSegmentedButton } from '../segmented-button-context';

export const useSegmentFoundation = (
  props: SegmentProps & React.HTMLProps<any>
) => {
  const [isSingleSelect, setIsSingleSelect] = useState(false);
  const [index, setIndex] = useState(-1);

  const contextApi = useSegmentedButton();

  const { foundation, ...elements } = useFoundation({
    props,
    elements: { rootEl: true },
    foundation: ({ rootEl, emit }) => {
      const getSegmentId = (): string | undefined => {
        return foundation.getSegmentId();
      };
      return new MDCSegmentedButtonSegmentFoundation({
        isSingleSelect: () => {
          return isSingleSelect;
        },
        getAttr: (attrName) => {
          return rootEl.getProp(attrName as any);
        },
        setAttr: (attrName, value) => {
          rootEl.ref?.setAttribute(attrName, value);
        },
        addClass: (className) => {
          rootEl.ref?.classList.add(className);
        },
        removeClass: (className) => {
          rootEl.ref?.classList.remove(className);
        },
        hasClass: (className) => {
          return rootEl.hasClass(className);
        },
        notifySelectedChange: (selected) => {
          emit<SegmentDetail>(
            'selected',
            {
              index: index,
              selected,
              segmentId: getSegmentId()
            },
            true /* shouldBubble */
          );
        },
        getRootBoundingClientRect: () => {
          return rootEl.ref?.getBoundingClientRect() ?? emptyClientRect;
        }
      });
    }
  });

  const { rootEl } = elements;

  const handleClick = () => foundation.handleClick();

  rootEl.ref?.addEventListener('click', handleClick);

  const segmentApi = useMemo<SegmentApi>(() => {
    return {
      setIsSingleSelect: (isSingleSelect) => setIsSingleSelect(isSingleSelect),
      index: rootEl.ref?.parentElement
        ? Array.from(rootEl.ref.parentElement.children).indexOf(rootEl.ref)
        : -1,
      setIndex: (index) => {
        setIndex(index);
        return index;
      },
      isSelected: () => foundation.isSelected(),
      setSelected: () => foundation.setSelected(),
      setUnselected: () => foundation.setUnselected(),
      getSegmentId: () => foundation.getSegmentId()
    };
  }, [foundation, rootEl.ref]);

  useEffect(() => {
    contextApi.registerSegment(segmentApi);

    return () => {
      contextApi.unregisterSegment(segmentApi);
    };
  }, [contextApi, segmentApi]);

  return {
    isSingleSelect,
    ...elements
  };
};
