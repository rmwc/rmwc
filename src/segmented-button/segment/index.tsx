import * as RMWC from '@rmwc/types';
import React from 'react';
import { withRipple } from '@rmwc/ripple';
import { Tag, useClassNames, createComponent } from '@rmwc/base';
import { useIsWrappedTouchTarget } from '@rmwc/touch-target';
import { useSegmentFoundation } from './foundation';

export interface SegmentProps extends RMWC.WithRippleProps {
  /** Make the button disabled */
  disabled?: boolean;
  /** Content specified as a label prop. */
  label?: React.ReactNode | string;
  /** Makes the button more touch friendly. This will automatically be set true if used inside of TouchTargetWrapper.*/
  touch?: boolean;
  /** Content specified as children. */
  children?: React.ReactNode;
  /** An Icon for the Button */
  icon?: React.ReactNode;
  /** A trailing icon for the Button */
  trailingIcon?: RMWC.IconPropT;
  /** Determines whether the button has been selected. */
  selected?: boolean;
}

export interface SegmentApi {
  setIsSingleSelect(isSingleSelect: boolean): void;
  index: number;
  setIndex(index: number): void;
  isSelected(): boolean;
  setSelected(): void;
  setUnselected(): void;
  getSegmentId(): string | undefined;
}

export type SegmentHTMLProps = RMWC.HTMLProps<HTMLButtonElement>;

export const Segment: RMWC.ComponentType<
  SegmentProps,
  SegmentHTMLProps,
  'button'
> = withRipple({
  surface: false
})(
  createComponent<SegmentProps, SegmentHTMLProps>(function Segment(props, ref) {
    const { children, label, icon, selected, touch, ...rest } = props;

    const { rootEl, isSingleSelect } = useSegmentFoundation(props);

    const isTouch = useIsWrappedTouchTarget(touch);

    const className = useClassNames(props, [
      'mdc-segmented-button__segment',
      {
        'mdc-segmented-button__segment--selected': selected,
        'mdc-segmented-button--touch': isTouch
      }
    ]);

    const selectedAccesibilityProps = isSingleSelect
      ? {
          'aria-checked': selected,
          role: 'radio'
        }
      : {
          'aria-pressed': selected
        };

    return (
      <Tag
        {...rest}
        tag="button"
        element={rootEl}
        className={className}
        ref={ref}
        {...selectedAccesibilityProps}
      >
        <div className="mdc-segmented-button__ripple"></div>
        {icon && (
          <i className="material-icons mdc-segmented-button__icon">{icon}</i>
        )}
        <div className="mdc-segmented-button__label">{label ?? children}</div>
      </Tag>
    );
  })
);
