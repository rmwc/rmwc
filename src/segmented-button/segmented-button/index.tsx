import * as RMWC from '@rmwc/types';
import React, { useRef } from 'react';

import { withRipple } from '@rmwc/ripple';
import { Tag, useClassNames, createComponent } from '@rmwc/base';
import { useSegmentedButtonFoundation } from './foundation';
import {
  SegmentedButtonContextT,
  SegmentedButtonProvider
} from '../segmented-button-context';

/*********************************************************************
 * Segmented Button
 *********************************************************************/

export type SegmentedButtonChangeEventT = RMWC.CustomEventT<{
  index: number;
}>;

export interface SegmentedButtonProps {
  /** Determines whether the user is able to select multiple or a single button at a time. Default to multiple. */
  selectType?: 'multiple' | 'single';
}

export type SegmentedButtonPropsHTMLProps = RMWC.HTMLProps<HTMLDivElement>;

export const SegmentedButton: RMWC.ComponentType<
  SegmentedButtonProps,
  SegmentedButtonPropsHTMLProps,
  'div'
> = withRipple({
  surface: false
})(
  createComponent<SegmentedButtonProps, SegmentedButtonPropsHTMLProps>(
    function SegmentedButton(props, ref) {
      const { children, selectType = 'multiple', ...rest } = props;

      const { rootEl, registerSegment, unregisterSegment } =
        useSegmentedButtonFoundation(props);

      const className = useClassNames(props, [
        'mdc-segmented-button',
        {
          'mdc-segmented-button--single-select': selectType === 'single'
        }
      ]);

      const contextApi = useRef<SegmentedButtonContextT>({
        registerSegment,
        unregisterSegment,
        selectType: selectType
      });

      return (
        <SegmentedButtonProvider value={contextApi.current}>
          <Tag
            {...rest}
            element={rootEl}
            tag="div"
            className={className}
            role={selectType === 'multiple' ? 'group' : 'radiogroup'}
            ref={ref}
          >
            {children}
          </Tag>
        </SegmentedButtonProvider>
      );
    }
  )
);
