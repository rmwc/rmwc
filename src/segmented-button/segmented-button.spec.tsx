import React from 'react';
import { render, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { SegmentedButton } from './segmented-button';
import { Segment } from './segment';

describe('Segmented Button', () => {
  it('renders', () => {
    const { asFragment } = render(
      <SegmentedButton selectType="single">
        <Segment icon="favorite" value="cookies" />
        <Segment label="Button" value="pizza" />
        <Segment icon="favorite" label="Button" value="icecream" />
      </SegmentedButton>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('handles onClick', async () => {
    const onClick = jest.fn();
    render(
      <SegmentedButton selectType="single">
        <Segment label="Cookies" value="cookies" onClick={onClick} />
      </SegmentedButton>
    );
    userEvent.click(screen.getByText('Cookies'));

    await waitFor(() => expect(onClick).toHaveBeenCalledTimes(1));
  });

  it('can be selected', () => {
    render(
      <SegmentedButton selectType="single">
        <Segment label="Cookies" value="cookies" selected />
      </SegmentedButton>
    );
    expect(screen.getByText('Cookies').parentElement).toHaveClass(
      'mdc-segmented-button__segment--selected'
    );
  });

  it('can have an icon', () => {
    const { asFragment } = render(
      <SegmentedButton>
        <Segment icon="favorite">Segment</Segment>
      </SegmentedButton>
    );
    expect(screen.getByText('favorite')).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  it('can have custom classnames', () => {
    render(
      <SegmentedButton>
        <Segment className={'my-custom-classname'}>Cookies</Segment>
      </SegmentedButton>
    );
    expect(screen.getByText('Cookies').parentElement).toHaveClass(
      'my-custom-classname'
    );
  });
});
