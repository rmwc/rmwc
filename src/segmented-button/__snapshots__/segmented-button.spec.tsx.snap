// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Segmented Button can have an icon 1`] = `
<DocumentFragment>
  <div
    class="mdc-segmented-button"
    role="group"
  >
    <button
      class="mdc-segmented-button__segment"
    >
      <div
        class="mdc-segmented-button__ripple"
      />
      <i
        class="material-icons mdc-segmented-button__icon"
      >
        favorite
      </i>
      <div
        class="mdc-segmented-button__label"
      >
        Segment
      </div>
    </button>
  </div>
</DocumentFragment>
`;

exports[`Segmented Button renders 1`] = `
<DocumentFragment>
  <div
    class="mdc-segmented-button mdc-segmented-button--single-select"
    role="radiogroup"
  >
    <button
      class="mdc-segmented-button__segment"
      role="radio"
      value="cookies"
    >
      <div
        class="mdc-segmented-button__ripple"
      />
      <i
        class="material-icons mdc-segmented-button__icon"
      >
        favorite
      </i>
      <div
        class="mdc-segmented-button__label"
      />
    </button>
    <button
      class="mdc-segmented-button__segment"
      role="radio"
      value="pizza"
    >
      <div
        class="mdc-segmented-button__ripple"
      />
      <div
        class="mdc-segmented-button__label"
      >
        Button
      </div>
    </button>
    <button
      class="mdc-segmented-button__segment"
      role="radio"
      value="icecream"
    >
      <div
        class="mdc-segmented-button__ripple"
      />
      <i
        class="material-icons mdc-segmented-button__icon"
      >
        favorite
      </i>
      <div
        class="mdc-segmented-button__label"
      >
        Button
      </div>
    </button>
  </div>
</DocumentFragment>
`;
